akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala

```diff
@@ -0,0 +1,320 @@
+/**
+ * Copyright (C) 2009-2014 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.io
+
+import java.net.InetSocketAddress
+import java.util.concurrent.atomic.AtomicBoolean
+
+import akka.actor.{ ActorRef, Terminated }
+import akka.dispatch.ExecutionContexts
+import akka.io.Inet.SocketOption
+import akka.io.Tcp
+import akka.io.Tcp._
+import akka.stream._
+import akka.stream.impl.ReactiveStreamsCompliance
+import akka.stream.impl.fusing.GraphStages.Detacher
+import akka.stream.scaladsl.Tcp.{ OutgoingConnection, ServerBinding }
+import akka.stream.scaladsl.{ BidiFlow, Flow, Tcp ⇒ StreamTcp }
+import akka.stream.stage.GraphStageLogic.StageActorRef
+import akka.stream.stage._
+import akka.util.ByteString
+
+import scala.collection.immutable
+import scala.concurrent.duration.{ Duration, FiniteDuration }
+import scala.concurrent.{ Future, Promise }
+
+/**
+ * INTERNAL API
+ */
+private[stream] class ConnectionSourceStage(val tcpManager: ActorRef,
+                                            val endpoint: InetSocketAddress,
+                                            val backlog: Int,
+                                            val options: immutable.Traversable[SocketOption],
+                                            val halfClose: Boolean,
+                                            val idleTimeout: Duration)
+  extends GraphStageWithMaterializedValue[SourceShape[StreamTcp.IncomingConnection], Future[StreamTcp.ServerBinding]] {
+
+  val out: Outlet[StreamTcp.IncomingConnection] = Outlet("IncomingConnections.out")
+  val shape: SourceShape[StreamTcp.IncomingConnection] = SourceShape(out)
+
+  private val BindTimer = "BindTimer"
+
+  // TODO: Timeout on bind
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[ServerBinding]) = {
+    val bindingPromise = Promise[ServerBinding]
+
+    val logic = new TimerGraphStageLogic(shape) {
+      implicit var self: StageActorRef = _
```

Comment:
Once #19015 is fixed there is no difference between constructor and `preStart` anymore.